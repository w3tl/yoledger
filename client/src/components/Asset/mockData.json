[{
  "__typename": "Account",
  "id": "1",
  "name": "Bank Card",
  "balance": 0
}, {
  "__typename": "Account",
  "id": "2",
  "name": "Cash",
  "balance": 10
}, {
  "__typename": "Account",
  "id": "3",
  "name": "Bank Card 2",
  "balance": 150
}]
