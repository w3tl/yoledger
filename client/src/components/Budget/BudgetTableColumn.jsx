import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import BudgetCell from './BudgetCell';
import { budgetPropType } from './propTypes';
import { accountPropType } from '../propTypes';
import withData from './BudgetTableColumnHOC';

function BudgetTableColumn({
  date, accounts, budget, onUpdate,
}) {
  return (
    <div>
      <div style={{ height: 30 }}>
        {moment(date).format('D MMM')}
      </div>
      {accounts.map(({ name }) => {
        const allocation = budget.find(b => b.account.name === name) || { amount: 0 };
        return (
          <BudgetCell key={`${date}_${name}`} amount={allocation.amount} onUpdate={onUpdate({ date, account: name })} />
        );
      })}
    </div>
  );
}

BudgetTableColumn.propTypes = {
  date: PropTypes.string.isRequired,
  accounts: PropTypes.arrayOf(accountPropType),
  budget: PropTypes.arrayOf(budgetPropType),
  onUpdate: PropTypes.func,
};

BudgetTableColumn.defaultProps = {
  accounts: [],
  budget: [],
  onUpdate: () => () => {},
};

export { BudgetTableColumn as BudgetTableColumnComponent };
export default withData(BudgetTableColumn);
