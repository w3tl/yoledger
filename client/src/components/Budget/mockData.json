{
  "__typename": "BudgetPlan",
  "periods": ["2018-06-10", "2018-06-25", "2018-07-10", "2018-07-25", "2018-08-10", "2018-08-25"],
  "accounts": [{
    "__typename": "Account",
    "id": "1",
    "name": "Food"
  }, {
    "__typename": "Account",
    "id": "2",
    "name": "Train"
  }],
  "budgets": [
    [{
      "id": "Period1Train",
      "account": {
        "name": "Train"
      },
      "amount": 11,
      "balance": 0
    }],
    [{
      "id": "Period2Train",
      "account": {
        "__typename": "Account",
        "id": "1",
        "name": "Food"
      },
      "amount": 12,
      "balance": 0
    }, {
      "id": "Period3Train",
      "account": {
        "name": "Train"
      },
      "amount": 13,
      "balance": 0
    }],
    [{
      "id": "Period4Train",
      "account": {
        "__typename": "Account",
        "id": "1",
        "name": "Food"
      },
      "amount": 14,
      "balance": 0
    }],
    [{
      "id": "Period5Train",
      "account": {
        "__typename": "Account",
        "id": "1",
        "name": "Food"
      },
      "amount": 16,
      "balance": 0
    }, {
      "id": "Period6Train",
      "account": {
        "name": "Train"
      },
      "amount": 17,
      "balance": 0
    }],
    [{
      "id": "Period7Train",
      "account": {
        "name": "Train"
      },
      "amount": 19,
      "balance": 0
    }],
    [{
      "id": "Period8Train",
      "account": {
        "__typename": "Account",
        "id": "1",
        "name": "Food"
      },
      "amount": 20,
      "balance": 0
    }, {
      "id": "Period9Train",
      "account": {
        "name": "Train"
      },
      "amount": 21,
      "balance": 0
    }]
  ]
}
