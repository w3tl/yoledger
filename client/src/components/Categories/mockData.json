{
  "expenses": [{
    "id": "21",
    "name": "Food",
    "type": "EXPENSE"
  }, {
    "id": "22",
    "name": "Beer",
    "type": "EXPENSE"
  }, {
    "id": "23",
    "name": "Train",
    "type": "EXPENSE"
  }],
  "incomes": [{
    "id": "31",
    "name": "Salary",
    "type": "INCOME"
  }, {
    "id": "32",
    "name": "Debts",
    "type": "INCOME"
  }, {
    "id": "33",
    "name": "Cashback",
    "type": "INCOME"
  }]
}
