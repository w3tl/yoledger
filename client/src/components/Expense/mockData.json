[{
  "id": "21",
  "name": "Food"
}, {
  "id": "22",
  "name": "Beer"
}, {
  "id": "23",
  "name": "Train"
}]
