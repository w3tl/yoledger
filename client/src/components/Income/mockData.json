[{
  "id": "31",
  "name": "Salary"
}, {
  "id": "32",
  "name": "Debts"
}, {
  "id": "33",
  "name": "Dividends"
}]
