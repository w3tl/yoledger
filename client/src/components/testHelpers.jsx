import React from 'react';
import { shallow, mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

export const withProvider = Component => props => (
  <MockedProvider {...props}>
    <Component />
  </MockedProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return fixedDate;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
