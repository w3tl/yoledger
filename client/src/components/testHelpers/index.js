import React from 'react';
import { shallow, mount } from 'enzyme';
import { ApolloProvider } from 'react-apollo';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { makeExecutableSchema } from 'graphql-tools';
import typeDefs from './typeDefs';
import resolvers from './resolvers';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

const executableSchema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

const cache = new InMemoryCache();
const link = new SchemaLink({ schema: executableSchema });
export const client = new ApolloClient({
  link,
  cache,
});

export const withProvider = Component => props => (
  <ApolloProvider client={client}>
    <Component client={client} />
  </ApolloProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  const date = new Date(fixedDate);
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return date;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
