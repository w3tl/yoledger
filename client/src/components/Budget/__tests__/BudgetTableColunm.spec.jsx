import React from 'react';
import { mount } from 'enzyme';
import { BudgetTableColumnComponent } from '../BudgetTableColumn';
import { periods, accounts, budgets } from '../mockData';
import { testRenderWithoutError } from '../../testHelpers';

describe('BudgetTableColumn', () => {
  testRenderWithoutError(
    <BudgetTableColumnComponent date={periods[0]} accounts={accounts} budget={budgets[0]} />,
  );

  it('should handle update cell', () => {
    const onUpdate = jest.fn();
    const wrapper = mount(
      <BudgetTableColumnComponent
        date={periods[0]}
        accounts={accounts}
        budget={budgets[0]}
        onUpdate={onUpdate}
      />,
    );

    wrapper.find('BudgetCell').first().find('input').simulate('blur');
    expect(onUpdate.mock.calls).toMatchSnapshot();
  });
});
