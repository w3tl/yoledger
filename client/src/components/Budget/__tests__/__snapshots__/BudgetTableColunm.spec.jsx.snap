// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BudgetTableColumn renders without error 1`] = `
<div>
  <div
    style={
      Object {
        "height": 30,
      }
    }
  >
    10 Jun
  </div>
  <BudgetCell
    amount={0}
    key="2018-06-10_Food"
    onUpdate={[Function]}
  />
  <BudgetCell
    amount={11}
    key="2018-06-10_Train"
    onUpdate={[Function]}
  />
</div>
`;

exports[`BudgetTableColumn should handle update cell 1`] = `
Array [
  Array [
    Object {
      "account": "Food",
      "date": "2018-06-10",
    },
  ],
  Array [
    Object {
      "account": "Train",
      "date": "2018-06-10",
    },
  ],
]
`;
